python/delta_sharing/rest_client.py

```diff
@@ -180,3 +190,30 @@ def _request_internal(self, request, target: str, data: Optional[Dict[str, Any]]
                 collections.deque(lines, maxlen=0)
         finally:
             response.close()
+
+    def _retry_with_exponential_backoff(self, func):
+        times_retried = 0
+        sleep_ms = 100
+        while True:
+            times_retried += 1
+            try:
+                return func()
+            except Exception as e:
+                if self._should_retry(e) and times_retried <= self._num_retries:
+                    logging.info(f"Sleeping {sleep_ms} to retry because of error {e}")
+                    self._sleeper(sleep_ms)
+                    sleep_ms *= 2
+                else:
+                    raise e
+
+    def _should_retry(self, error):
+        if isinstance(error, HTTPError):
+            error_code = error.response.status_code
+            if error_code == 429:
+                return True
+            elif 500 <= error_code <= 600:
+                return True
+            else:
+                return False
```

Comment:
I'm not sure what 
> case _: InterruptedException => false
      case _: InterruptedIOException => false
      case _: IOException => true
     
map to, so I just did HTTPError for now.

Reply:
I'm not sure there is a corresponding error from `requests` module.
I guess it wraps such errors in `HTTPError`.

Reply:
What's the error when the server is down? Could you try it? It would be great we can still retry so that even if we are restarting the server, the client can still survive.